### ISSUE TITLE
multiple inference on tensorrt

### ISSUE THREAD

**Issue by hailcpy** [jkjung-avt/tensorrt_demos#203]
Hey i am trying to run two tensorrt models on my jetson nano. One is the mtcnn model from your repo. The other is the arcface model for face recognition which i converted into tensorrt using python tensorrt api. I can run these models separately but on running them together i run into some weird errors which you had said is because of two cuda context. I saw your code on trtssd async but you had only one tensorrt model. How can i replicate that for two models. Also i have difficulty understanding how you have implemented the tensorrt mtcnn model and what is pytrt so i am just importing TrtMtcnn from utils.mtcnn and then just passing it image to get bounding box and landmarks. I don't know if i should first do cfx = cuda.device(0).make_context() which i am doing for the other model. Please help. I am new to tensorrt.

**Comment by jkjung-avt** [comment]
> Also i have difficulty understanding how you have implemented the tensorrt mtcnn model and what is pytrt so i am just importing TrtMtcnn from utils.mtcnn and then just passing it image to get bounding box and landmarks.

This is correct.  For TrtMtcnn, you don't need to create any additional CUDA context.

> I can run these models separately but on running them together i run into some weird errors which you had said is because of two cuda context.

Please be specific.  What error message do you see when running the 2 TensorRT engines together?

**Comment by hailcpy** [comment]
Sorry. Here is the error i get:
Cuda error in file src/implicit_gemm.cu at line 1224: invalid resource handle
[TensorRT] ERROR: ../rtExt/cuda/customWinogradConvActRunner.cpp (317) - TRTInternal Error in execute: 400 (Failure to run convolution)
[TensorRT] ERROR: FAILED_EXECUTION: std::exception
If i run both separately they work fine.

**Comment by hailcpy** [comment]
This is the class for the other tensorrt model i am using along with mtcnn. I call the function infer to do inference. First the mtcnn gives bounding box and landmarks which i then use to crop and align the face image and then feed the modified image to this model which recognizes the face.
class arcfaceTRT(object):
	'''
	description: A arcface class that warps TensorRT ops, preprocess and postprocess ops.
	'''
	def __init__(self, engine_file_path):
		# Create a Context on this device,
		self.cfx = cuda.Device(0).make_context()
		stream = cuda.Stream()
		TRT_LOGGER = trt.Logger(trt.Logger.INFO)
		runtime = trt.Runtime(TRT_LOGGER)

		# Deserialize the engine from file
		with open(engine_file_path, "rb") as f:
			engine = runtime.deserialize_cuda_engine(f.read())
		context = engine.create_execution_context()

		host_inputs = []
		cuda_inputs = []
		host_outputs = []
		cuda_outputs = []
		bindings = []

		for binding in engine:
			size = trt.volume(engine.get_binding_shape(
				binding)) * engine.max_batch_size
			dtype = trt.nptype(engine.get_binding_dtype(binding))
			# Allocate host and device buffers
			host_mem = cuda.pagelocked_empty(size, dtype)
			cuda_mem = cuda.mem_alloc(host_mem.nbytes)
			# Append the device buffer to device bindings.
			bindings.append(int(cuda_mem))
			# Append to the appropriate list.
			if engine.binding_is_input(binding):
				host_inputs.append(host_mem)
				cuda_inputs.append(cuda_mem)
			else:
				host_outputs.append(host_mem)
				cuda_outputs.append(cuda_mem)

		# Store
		self.stream = stream
		self.context = context
		self.engine = engine
		self.host_inputs = host_inputs
		self.cuda_inputs = cuda_inputs
		self.host_outputs = host_outputs
		self.cuda_outputs = cuda_outputs
		self.bindings = bindings

	def infer(self, input_image_path):
		threading.Thread.__init__(self)
		# Make self the active context, pushing it on top of the context stack. 
		self.cfx.push()
		# Restore
		stream = self.stream
		context = self.context
		engine = self.engine
		host_inputs = self.host_inputs
		cuda_inputs = self.cuda_inputs
		host_outputs = self.host_outputs
		cuda_outputs = self.cuda_outputs
		bindings = self.bindings
		# Do image preprocess
		input_image, image_raw, origin_h, origin_w = self.preprocess_image(
			input_image_path)
		# Copy input image to host buffer
		np.copyto(host_inputs[0], input_image.ravel())
		# Transfer input data  to the GPU.
		cuda.memcpy_htod_async(cuda_inputs[0], host_inputs[0], stream)
		# Run inference.
		context.execute_async(bindings=bindings, stream_handle=stream.handle)
		# Transfer predictions back from the GPU.
		cuda.memcpy_dtoh_async(host_outputs[0], cuda_outputs[0], stream)
		# Synchronize the stream
		stream.synchronize()
		# Remove any context from the top of the context stack, deactivating it.
		self.cfx.pop()
		del self.cfx
		# Here we use the first row of output in that batch_size = 1
		output = host_outputs[0]
		return output

	def destory(self):
		# Remove any context from the top of the context stack, deactivating it.
		self.cfx.pop()

	def preprocess_image(self, input_image_path):
		'''
		description: Read an image from image path, convert it to RGB,
						resize and pad it to target size, normalize to [0,1],
						transform to NCHW format.
		param:
			input_image_path: str, image path
		return:
			image:  the processed image
			image_raw: the original image
			h: original height
			w: original width
		'''
		image_raw = input_image_path
		h, w, c = image_raw.shape
		image = cv2.cvtColor(image_raw, cv2.COLOR_BGR2RGB)
		# Calculate widht and height and paddings
		r_w = INPUT_W / w
		r_h = INPUT_H / h
		if r_h > r_w:
			tw = INPUT_W
			th = int(r_w * h)
			tx = 0
			ty = int((INPUT_H - th) / 2)
		else:
			tw = int(r_h * w)
			th = INPUT_H
			tx = int((INPUT_W - tw) / 2)
			ty = 0
		# Resize the image with long side while maintaining ratio
		image = cv2.resize(image, (tw, th))
		# Pad the short side with (128,128,128)
		image = cv2.copyMakeBorder(
			image, ty, ty, tx, tx, cv2.BORDER_CONSTANT, (128, 128, 128))
		image = image.astype(np.float32)
		# Normalize to [0,1]
		image -= 127.5
		image = image*0.0078125
		# HWC to CHW format:
		image = np.transpose(image, [2, 0, 1])
		# CHW to NCHW format
		image = np.expand_dims(image, axis=0)
		# Convert the image to row-major order, also known as "C order":
		image = np.ascontiguousarray(image)
		return image, image_raw, h, w

**Comment by jkjung-avt** [comment]
I think it would work if you just **remove** cuda context related code in your "arcfaceTRT" class.

```
class arcfaceTRT(object):
    def __init__(self, engine_file_path):
        self.cfx = cuda.Device(0).make_context()   # REMOVE THIS LINE
        ......

    def infer(self, input_image_path):
        threading.Thread.__init__(self)    # REMOVE THIS LINE
        # Make self the active context, pushing it on top of the context stack. 
        self.cfx.push()                            # REMOVE THIS LINE
        ......
        self.cfx.pop()                             # REMOVE THIS LINE
        del self.cfx                               # REMOVE THIS LINE
        .....

    def destory(self):
        # Remove any context from the top of the context stack, deactivating it.
        self.cfx.pop()                             # REMOVE THIS LINE (REAPLCE with `pass`)
        ......
```

**Comment by hailcpy** [comment]
It solved the problem. Thank you so much. Can you please tell me what was the problem. Thanks.

**Comment by jkjung-avt** [comment]
The short answer is: for a single-threaded python program, you don't need to create additional cuda contexts to run multiple TensorRT engines.  (In this case, all TensorRT engines would be using the "default context" created by [pycuda.autoinit](https://documen.tician.de/pycuda/util.html#module-pycuda.autoinit).)

To fully understand it, you'll need to understand low-level CUDA programming and read pycuda documentation...

**Comment by hailcpy** [comment]
In this code the libraries i was using are as follows:
import ctypes
import os
import random
import sys
import threading
import time

import cv2
import numpy as np
import pycuda.autoinit
import pycuda.driver as cuda
import tensorrt as trt
import torch
import torchvision
import sklearn.preprocessing

if i add "import mxnet as mx" i get "PyCUDA ERROR: The context stack was not empty upon module cleanup."
I need to use an mxnet model along with these two models. But i keep getting this error. Please help. I think the problem is caused when pycuda.autoint and mxnet are used together

**Comment by jkjung-avt** [comment]
With the latest code in the repo, I think this could be fix the problem: https://github.com/jkjung-avt/tensorrt_demos/issues/265#issuecomment-712006217

Please give it a try.

**Comment by hailcpy** [comment]
Hey i checked your latest code. The problem isn't with your code. I can use it fine with mxnet model. The problem is with the arcface tensorrt model i was using. I was using pycuda.autoinit in that. If i just add the line "import mxnet as mx" i get "PyCUDA ERROR: The context stack was not empty upon module cleanup."  I think it is because of pycuda.autoinit. 
The code of arcface tensorrt is posted above. Can you please tell me what is the issue and how can i solve it.

**Comment by jkjung-avt** [comment]
Please refer to the following code snippet in "utils/yolo_with_plugins.py".  You could try to call `push()` with `pycuda.autoinit.context` before inferencing with the TensorRT engine, and call `pop()` afterwards.

https://github.com/jkjung-avt/tensorrt_demos/blob/e039e0824876d46443aa19e9ad7cf8f7723c713e/utils/yolo_with_plugins.py#L277-L286

**Comment by jkjung-avt** [comment]
> The problem isn't with your code. I can use it fine with mxnet model.

Closing this since the remaining issue is not related to code in this repo.
